cmake/tribits/core/package_arch/TribitsWriteClientExportFiles.cmake

```diff
@@ -925,3 +933,21 @@ include(\"${${TRIBITS_PACKAGE}_BINARY_DIR}/${TRIBITS_PACKAGE}Config.cmake\")")
 
 ENDFUNCTION()
 
+
+MACRO(TRIBITS_SET_COMPILER_VAR_FOR_CONFIG_FILE LANG FOR_DIR)
+  IF (NOT "${CMAKE_${LANG}_COMPILER_FOR_CONFIG_FILE_${FOR_DIR}}" STREQUAL "")
+    SET(CMAKE_${LANG}_COMPILER_FOR_CONFIG_FILE
+      "${CMAKE_${LANG}_COMPILER_FOR_CONFIG_FILE_${FOR_DIR}}")
+  ELSE()
+    SET(CMAKE_${LANG}_COMPILER_FOR_CONFIG_FILE
+      "${CMAKE_${LANG}_COMPILER}")
+  ENDIF()
+  #MESSAGE("${FOR_DIR}: CMAKE_${LANG}_COMPILER_FOR_CONFIG_FILE='${CMAKE_${LANG}_COMPILER_FOR_CONFIG_FILE}'")
+ENDMACRO()
+
+
+MACRO(TRIBITS_SET_COMPILER_VARS_FOR_CONFIG_FILE FOR_DIR)
```

Comment:
This needs to be defined above https://github.com/trilinos/Trilinos/pull/7642/files#diff-9a261b6fb235aff56c7ea49c6c9c1063R435, no?

Reply:
No, CMake allows you to call functions and macros before you define them (just like Python does).  CMake seems to read the entire program into memory before it executes any commands contained in that program.

Reply:
A simple CMakeLists.txt shows different behavior:
```cmake
test_print("1")

macro(test_print arg)
  message("test print ${arg}")
endmacro()

test_print("2")
```

```bash
$ cmake ..
CMake Warning (dev) in CMakeLists.txt:
  No project() command is present.  The top-level CMakeLists.txt file must
  contain a literal, direct call to the project() command.  Add a line of
  code such as

    project(ProjectName)

  near the top of the file, but after cmake_minimum_required().

  CMake is pretending there is a "project(Project)" command on the first
  line.
This warning is for project developers.  Use -Wno-dev to suppress it.
<snip>
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:1 (test_print):
  Unknown CMake command "test_print".
<snip>
```

Reply:
Not sure what is going on here.  But to test that file properly, you have to run it with `cmake -P CMakeLists.txt` since it is not really a CMake project file.  It is just a CMake script.

Reply:
I seem to be getting the same error:
```
$ cmake -P CMakeLists.txt 
CMake Error at CMakeLists.txt:1 (test_print):
  Unknown CMake command "test_print".
```

Can you confirm that we're not seeing: Unknown CMake command "TRIBITS_SET_COMPILER_VARS_FOR_CONFIG_FILE" with these changes?

Reply:
Hum, I will have to look into this.  That macro does not seem to be getting called in functions/macros in that file.

Reply:
@e10harvey, if you comment that macro out, TriBITS nor Trilinos will even configure.  I just tried this with diff:

```
[rabartl@crf450 TriBITS (master)]$ git diff
diff --git a/tribits/core/package_arch/TribitsWriteClientExportFiles.cmake b/tribits/core/package_arch/TribitsWriteClientExportFiles.cmake
index 02d003a..7e85a95 100644
--- a/tribits/core/package_arch/TribitsWriteClientExportFiles.cmake
+++ b/tribits/core/package_arch/TribitsWriteClientExportFiles.cmake
@@ -946,8 +946,8 @@ MACRO(TRIBITS_SET_COMPILER_VAR_FOR_CONFIG_FILE LANG FOR_DIR)
 ENDMACRO()
 
 
-MACRO(TRIBITS_SET_COMPILER_VARS_FOR_CONFIG_FILE FOR_DIR)
-  TRIBITS_SET_COMPILER_VAR_FOR_CONFIG_FILE(CXX ${FOR_DIR})
-  TRIBITS_SET_COMPILER_VAR_FOR_CONFIG_FILE(C ${FOR_DIR})
-  TRIBITS_SET_COMPILER_VAR_FOR_CONFIG_FILE(Fortran ${FOR_DIR})
-ENDMACRO()
+#MACRO(TRIBITS_SET_COMPILER_VARS_FOR_CONFIG_FILE FOR_DIR)
+#  TRIBITS_SET_COMPILER_VAR_FOR_CONFIG_FILE(CXX ${FOR_DIR})
+#  TRIBITS_SET_COMPILER_VAR_FOR_CONFIG_FILE(C ${FOR_DIR})
+#  TRIBITS_SET_COMPILER_VAR_FOR_CONFIG_FILE(Fortran ${FOR_DIR})
+#ENDMACRO()
```

and I get the TriBITS configure error:

```
CMake Error at tribits/core/package_arch/TribitsWriteClientExportFiles.cmake:548 (TRIBITS_SET_COMPILER_VARS_FOR_CONFIG_FILE):
  Unknown CMake command "TRIBITS_SET_COMPILER_VARS_FOR_CONFIG_FILE".
Call Stack (most recent call first):
  tribits/core/package_arch/TribitsWriteClientExportFiles.cmake:659 (TRIBITS_WRITE_FLEXIBLE_PACKAGE_CLIENT_EXPORT_FILES)
  tribits/core/package_arch/TribitsPackageMacros.cmake:716 (TRIBITS_WRITE_PACKAGE_CLIENT_EXPORT_FILES)
  tribits/core/package_arch/TribitsPackageMacros.cmake:819 (TRIBITS_PACKAGE_POSTPROCESS_COMMON)
  CMakeLists.txt:22 (TRIBITS_PACKAGE_POSTPROCESS)
  tribits/core/package_arch/TribitsGlobalMacros.cmake:2664 (INCLUDE)
  tribits/core/package_arch/TribitsProjectImpl.cmake:257 (TRIBITS_CONFIGURE_ENABLED_PACKAGES)
  tribits/core/package_arch/TribitsProjectImpl.cmake:364 (TRIBITS_PROJECT_IMPL)
  CMakeLists.txt:17 (TRIBITS_PROJECT_ENABLE_ALL)


-- Configuring incomplete, errors occurred!
See also "/ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/TRIBITS_MPI_DEBUG/CMakeFiles/CMakeOutput.log".
make: *** [cmake_check_build_system] Error 1
```

This is proof that this macro is used.

Can you please approve this PR so it can merge?